src/library/scala/concurrent/Future.scala

```diff
@@ -413,8 +413,37 @@ trait Future[+T] extends Awaitable[T] {
    * @return        a `Future` with the result of the application of `f` to the results of `this` and `that`
    * @group Transformations
    */
-  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] =
-    flatMap(r1 => that.map(r2 => f(r1, r2)))(if (executor.isInstanceOf[BatchingExecutor]) executor else parasitic)
+  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
+    val sentinel = new Object()
```

Comment:
The sentinel is not needed if you use the logic I proposed in https://github.com/scala/scala/pull/9655/files#r645659226 and thus saves an allocation.